pylint/config/arguments_manager.py

```diff
@@ -0,0 +1,107 @@
+        """Namespace for all options."""
+
+        self._arg_parser = argparse.ArgumentParser(prog="pylint")
+        """The command line argument parser."""
+
+        self._argument_groups_dict: Dict[str, argparse._ArgumentGroup] = {}
+        """Dictionary of all the argument groups."""
+
+    def _register_options_provider(self, provider: "checkers.BaseChecker") -> None:
+        """Register an options provider and loads its defaults."""
+        # pylint: disable-next=fixme
+        # TODO: Do something own_group parameter (see OptionsManagerMixIn.register_options_provider)
+        for opt, optdict in provider.options:
+            argument = _convert_option_to_argument(opt, optdict)
+            self._add_arguments_to_parser(provider.name, argument)
+
+        # pylint: disable-next=fixme
+        # TODO: Do something with option provider priorities (see OptionsManagerMixIn.register_options_provider)
+        # pylint: disable-next=fixme
+        # TODO: Do something with option groups within optdicts (see OptionsManagerMixIn.register_options_provider)
+
+        # pylint: disable-next=fixme
+        # TODO: Investigate performance impact of loading default arguments on every call
+        self._load_default_argument_values()
+
+    def _add_arguments_to_parser(self, section: str, argument: _Argument) -> None:
+        """Iterates over all argument sections and add them to the parser object."""
+        try:
+            section_group = self._argument_groups_dict[section]
+        except KeyError:
+            section_group = self._arg_parser.add_argument_group(title=section)
+            self._argument_groups_dict[section] = section_group
+        self._add_parser_option(section_group, argument)
+
+    @staticmethod
+    def _add_parser_option(
+        section_group: argparse._ArgumentGroup, argument: _Argument
+    ) -> None:
+        """Add an argument."""
+        if argument.action == "store":
+            section_group.add_argument(
+                *argument.name,
+                action=argument.action,
+                default=argument.default,
+                type=argument.type,  # type: ignore[arg-type] # incorrect typing in typeshed
+                help=argument.help,
+                metavar=argument.metavar,
+                choices=argument.choices,
+            )
+        else:
+            raise UnrecognizedArgumentAction
+
+    def _load_default_argument_values(self) -> None:
+        """Loads the default values of all registered options."""
+        self.namespace = self._arg_parser.parse_args([], self.namespace)
+
+    def _parse_configuration_file(
+        self, config_parser: configparser.ConfigParser
+    ) -> None:
+        """Parse the arguments found in a configuration file into the namespace."""
+        arguments = []
+        for section in config_parser.sections():
+            for opt, value in config_parser.items(section):
+                arguments.extend([f"--{opt}", value])
```

Comment:
This is probably a nothing. But, since you know the parsing flow very well, I'll ask -- is there a chance `opt` is already in `arguments`? And if so, does `argparse` pick the right one later?

Reply:
Hm, I don't actually know how double arguments are picked up by the `config_parser`. Although I think that it will probably raise a `configparser.DoubleKeyError` or something. (`configparser` is not our own code).

The arguments in the namespace get overwritten if they get re-parsed. That's why the order in `config_initialization` is important.